test/e2e/conftest.py

```diff
@@ -213,78 +213,97 @@ def worker(
-        deadline_client = boto3.client("deadline")
-
-        worker = ec2_worker_type(
-            ec2_client=ec2_client,
-            s3_client=s3_client,
-            deadline_client=deadline_client,
-            bootstrap_bucket_name=bootstrap_resources.bootstrap_bucket_name,
-            ssm_client=ssm_client,
-            override_ami_id=ami_id,
-            subnet_id=subnet_id,
-            security_group_id=security_group_id,
-            instance_profile_name=bootstrap_resources.worker_instance_profile_name,
-            configuration=worker_config,
-            instance_type=instance_type,
-            instance_shutdown_behavior=instance_shutdown_behavior,
-        )
-
-    def stop_worker():
-        if request.session.testsfailed > 0:
-            if os.getenv("KEEP_WORKER_AFTER_FAILURE", "false").lower() == "true":
-                LOG.info("KEEP_WORKER_AFTER_FAILURE is set, not stopping worker")
-                return
+            assert subnet_id, "SUBNET_ID is required when deploying an EC2 worker"
+            assert security_group_id, "SECURITY_GROUP_ID is required when deploying an EC2 worker"
+
+            bootstrap_resources: BootstrapResources = request.getfixturevalue("bootstrap_resources")
+            assert (
+                bootstrap_resources.worker_instance_profile_name
+            ), "Worker instance profile is required when deploying an EC2 worker"
+
+            ec2_client = boto3.client("ec2")
+            s3_client = boto3.client("s3")
+            ssm_client = boto3.client("ssm")
+            deadline_client = boto3.client("deadline")
+
+            worker = ec2_worker_type(
+                ec2_client=ec2_client,
+                s3_client=s3_client,
+                deadline_client=deadline_client,
+                bootstrap_bucket_name=bootstrap_resources.bootstrap_bucket_name,
+                ssm_client=ssm_client,
+                override_ami_id=ami_id,
+                subnet_id=subnet_id,
+                security_group_id=security_group_id,
+                instance_profile_name=bootstrap_resources.worker_instance_profile_name,
+                configuration=worker_config,
+                instance_type=instance_type,
+                instance_shutdown_behavior=instance_shutdown_behavior,
+            )
 
         try:
-            worker.stop()
+            worker.start()
         except Exception as e:
-            LOG.exception(f"Error while stopping worker: {e}")
-            LOG.error(
-                "Failed to stop worker. Resources may be left over that need to be cleaned up manually."
-            )
+            LOG.exception(f"Failed to start worker: {e}")
+            LOG.info("Stopping worker because it failed to start")
+            stop_worker(request, worker)
             raise
 
+        all_created_workers.append(worker)
+        return worker
+
+    yield generate_worker
+
+    for worker in all_created_workers:
+        stop_worker(request, worker)
```

Comment:
This doesn't seem too correct. the request would be consistent across the session. So if any tests fail across the session, we don't shut down any of the workers?

Reply:
Changed in the new commit to make sure function worker are killed once the tests finish rather then wait until the end and got deleted with the session worker.